### PR TITLE
rook: add logCollector in storageCluster and update to rook v1.9.9

### PR DESCRIPTION
**Pull request by subhamkrai**
upgrade to rook v1.9.9

rook: add logCollector in storageCluster

adding logCollector in storageCluster which will have
fields like `enabled` to enable/disable logCollector,
`periodicity` interval after which logs will be rotated
and `maxLogSize` time after which logs will be rotated.
If both `periodicity` and `maxLogSize` what every comes
first will rotate the logs.

### CODE REVIEW

**Comment by subhamkrai** [comment]
1.
```
spec:
  logCollector:
    enabled: true
    #maxLogSize: 1G
    #periodicity: 1d
```
```
cat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.a.log {
    rotate 7
    daily
    maxsize 500M
    compress
    sharedscripts
    postrotate
        killall -q -1 ceph-mon ceph-mgr ceph-mds ceph-osd ceph-fuse radosgw rbd-mirror cephfs-mirror || pkill -1 -x "ceph-mon|ceph-mgr|ceph-mds|ceph-osd|ceph-fuse|radosgw|rbd-mirror|cephfs-mirror" || true
    endscript
    missingok
    notifempty
    su root ceph
}
```

2. 
```
logCollector:
    enabled: true
    #maxLogSize: 1G
    periodicity: 1h
```
```
cat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.b.log {
    rotate 7
    hourly
    maxsize 500M
    compress
    sharedscripts
    postrotate
        killall -q -1 ceph-mon ceph-mgr ceph-mds ceph-osd ceph-fuse radosgw rbd-mirror cephfs-mirror || pkill -1 -x "ceph-mon|ceph-mgr|ceph-mds|ceph-osd|ceph-fuse|radosgw|rbd-mirror|cephfs-mirror" || true
    endscript
    missingok
    notifempty
    su root ceph
}
```

3. 
```
logCollector:
    enabled: true
    maxLogSize: 1G
    periodicity: daily
```
```
cat /etc/logrotate.d/ceph p
/var/log/ceph/ceph-mon.a.log {
    rotate 7
    daily
    maxsize 1G
    compress
    sharedscripts
    postrotate
        killall -q -1 ceph-mon ceph-mgr ceph-mds ceph-osd ceph-fuse radosgw rbd-mirror cephfs-mirror || pkill -1 -x "ceph-mon|ceph-mgr|ceph-mds|ceph-osd|ceph-fuse|radosgw|rbd-mirror|cephfs-mirror" || true
    endscript
    missingok
    notifempty
    su root ceph
}
```

4.
```
logCollector:
    enabled: true
    maxLogSize: 700Mi
    periodicity: weekly
```
```
cat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.c.log {
    rotate 7
    weekly
    maxsize 734M
    compress
    sharedscriptscat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.c.log {
    rotate 7
    weekly
    maxsize 734M
    compress
    sharedscripts
```

5.
```
logCollector:
    enabled: true
    maxLogSize: 1Gi
    periodicity: monthly
```
```
 cat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.b.log {
    rotate 7
    monthly
    maxsize 1073M
    compress
    sharedscripts
    postrotate
        killall -q -1 ceph-mon ceph-mgr ceph-mds ceph-osd ceph-fuse radosgw rbd-mirror cephfs-mirror || pkill -1 -x "ceph-mon|ceph-mgr|ceph-mds|ceph-osd|ceph-fuse|radosgw|rbd-mirror|cephfs-mirror" || true
    endscript
    missingok
    notifempty
    su root ceph
} 
```

6.
```
logCollector:
    enabled: true
    maxLogSize: 800M
    periodicity: hourly
```

```
cat /etc/logrotate.d/ceph 
/var/log/ceph/ceph-mon.b.log {
    rotate 7
    hourly
    maxsize 800M
    compress
    sharedscripts
    postrotate
        killall -q -1 ceph-mon ceph-mgr ceph-mds ceph-osd ceph-fuse radosgw rbd-mirror cephfs-mirror || pkill -1 -x "ceph-mon|ceph-mgr|ceph-mds|ceph-osd|ceph-fuse|radosgw|rbd-mirror|cephfs-mirror" || true
    endscript
    missingok
    notifempty
    su root ceph
}
```

File: controllers/storagecluster/cephcluster.go
Comment (travisn) [inline controllers/storagecluster/cephcluster.go:413]: Could you specify the size with "500Mi" instead of the bytes?

File: controllers/storagecluster/cephcluster.go
Comment (travisn) [inline controllers/storagecluster/cephcluster.go:409]: If it's nullable, the pattern could be to first set the default, then adjust if settings are specified:
```suggestion
	cephCluster.Spec.LogCollector.MaxLogSize = resource.NewQuantity("500Mi", resource.BinarySI)
	if cephCluster.Spec.LogCollector != nil {
```

File: api/v1/storagecluster_types.go
Comment (travisn) [inline api/v1/storagecluster_types.go:96]: Let's make this nullable so we can detect if we should use defaults or not.
```suggestion
	LogCollector *rookCephv1.LogCollectorSpec `json:"logCollector,omitempty"`
```

File: vendor/modules.txt
Comment (travisn) [inline vendor/modules.txt:463]: The backports for this may be difficult since #1652 wasn't backported. For backport, we will need a more restricted update than the full `make deps-update` that was run for #1652.

**Review by travisn — CHANGES_REQUESTED** [review]

File: controllers/storagecluster/cephcluster.go
Comment (subhamkrai) [reply controllers/storagecluster/cephcluster.go:413]: resource.NewQuantity takes `int64` as a first arg and we have `resource.NewQuantity` only because we need the pointer return.

Comment (travisn) [reply controllers/storagecluster/cephcluster.go:413]: It's convenient that it returns the pointer directly, but I worry about accidentally setting the wrong value since there are so many zeros. How about `ParseQuantity("500Gi")`?

Comment (subhamkrai) [reply controllers/storagecluster/cephcluster.go:413]: done.

File: vendor/modules.txt
Comment (subhamkrai) [reply vendor/modules.txt:463]: right, it will give errors without the changes of #1652.

**Comment by subhamkrai** [comment]
Also, I have kept `hold` label since I would like to test again once rook downstream pr is merged(upstream is already merged) just to double sure.

**Comment by agarwal-mudit** [comment]
@subhamkrai this is required for release-4.11 as well, right?

**Comment by subhamkrai** [comment]
> @subhamkrai this is required for release-4.11 as well, right?

right, thanks for reminding.

**Comment by subhamkrai** [comment]
keeping /hold since rook backport is not merged yet

**Comment by subhamkrai** [comment]
/hold cancel 

rook bp are merged.

**Pull request merged**
